Create bdd-case-study (#81)
incubyte/incubyte.github.io: content/blog/bdd-case-study.md

+++
title = "Case-Study Behaviour Driven Development"
slug = "bdd-case-study"
date = 2022-07-18T02:32:46Z
image = ""
draft = false
authors = ["Incubyte"]
description = ""
tags = ["BDD","TDD", "Software Craftsmanship"]
categories = ["Playbook", "Software Craftsmanship"]
type = ""
+++

> ## Implementing Behavior Driven Deployment methodologies to get better Software out the door - a case study

Reviewer: Development, not Deployment. Also, check capitalization - don't think Software should be capitalized

> Over the years, the software development process has seen a massive change of pace. Through ideation, design, development, testing and implementation all the processes involve multiple teams. These groups have varied responsibilities and need to work in a cohesive manner to achieve the end goal.

Reviewer: ```suggestion
Over the years, the software development process has seen a massive change of pace. Through ideation, design, development, testing, and implementation, all processes involve multiple skills, which have traditionally been served through multiple teams. These groups have varied responsibilities and need to work in a cohesive manner to achieve the end goal.
```

> ## Bogged down by Waterfall:

Reviewer: ```suggestion
## Bogged down by Waterfall
```

Traditional development methodologies having linear processes don’t cut it these days. A frequently changing business landscape and evolving user requirements drive the need to deliver regular and incremental value in thin slices. To get to the point of getting a deliverable ready looks more like the diagram below with multiple points of communication rather than a straight line.

<img width="352" alt="Screenshot 2022-07-18 at 1 28 47 PM" src="https://user-images.githubusercontent.com/81680332/179468238-81b6a1c5-1b4c-4918-95cd-5d1223568831.png">

Fig 1: Communication between multiple distributed teams

Here are some challenges with traditional methodologies:
·	Vagueness in project requirements due to working in silos as per development process phase
·	Improper API documentation as the project grows
·	Treating QA testing as the last process before release
·	Lack of process automation and resulting slower release cycles

> ## Focus on behavior and not tools:

Reviewer: ```suggestion
## Focus on behavior and not tools
```

> ## Focus on behavior and not tools

Reviewer 2: I think it should be not THE tool

“A software system can best be designed if the testing is interlaced with the designing, instead of being used after the designing.”
-	Alan Perils

These days we’re seeing the terms TDD [Test Driven Development, link: https://www.geeksforgeeks.org/test-driven-development-tdd/] and BDD [Behavior Driven Development, link: https://www.techtarget.com/searchsoftwarequality/definition/Behavior-driven-development-BDD] being tossed about a lot, but why the fuss? Radical as it may seem, focus on these processes have become a necessity now if you consider the data from various surveys that show how developers spend almost 1/3rd of their time working to fix issues in their code instead of working on a new feature/enhancement. [source: https://www.businesswire.com/news/home/20210216005484/en/Rollbar-Research-Shows-That-Traditional-Error-Monitoring-Is-Missing-the-Mark]

> ## TDD helps circumvent these issues by:

Reviewer: ```suggestion
**Practices like TDD and BDD help circumvent these issues by:**
```

•	Forcing attention on software design from the get-go with a focus on modularity
•	Promoting an iterative development strategy
•	Providing a better scalability and maintainability opportunity
•	Removing the veil of uncertainties involving refactoring code

> Using the “shift testing/QA to the left” approach, BDD ensures the essence of what the software is supposed to do, is completely and evenly spread across all vital stakeholders, including QAs and PAs.

Reviewer: ```suggestion
Enabling a "fail fast" approach, BDD ensures that the essence of what the software is supposed to do is created in an executable format, bringing in not just automation, but also full clarity around expected behaviour from the get-go.
```
